pkg/runtime/framework/plugins/jobset/jobset_test.go

```diff
@@ -325,3 +330,109 @@ func TestJobSet(t *testing.T) {
 		})
 	}
 }
+
+func TestJobSetValidate(t *testing.T) {
+	cases := map[string]struct {
+		runtimeJobTemplate client.Object
+		runtimeInfo        *runtime.Info
+		oldObj             *trainer.TrainJob
+		newObj             *trainer.TrainJob
+		wantError          field.ErrorList
+		wantWarnings       admission.Warnings
+	}{
+		"no jobset runtime template": {
+			runtimeJobTemplate: &utiltesting.JobSetWrapper{},
+			newObj: utiltesting.MakeTrainJobWrapper(metav1.NamespaceDefault, "test").
+				Obj(),
+		},
+		"no initializer job": {
+			runtimeJobTemplate: utiltesting.MakeJobSetWrapper(metav1.NamespaceDefault, "test"),
+			newObj: utiltesting.MakeTrainJobWrapper(metav1.NamespaceDefault, "test").
+				Obj(),
+		},
+		"no dataset intializer job": {
+			runtimeJobTemplate: utiltesting.MakeJobSetWrapper(metav1.NamespaceDefault, "test"),
+			newObj: utiltesting.MakeTrainJobWrapper(metav1.NamespaceDefault, "test").
+				Initializer(&trainer.Initializer{Dataset: nil}).
+				Obj(),
+			wantError: field.ErrorList{
+				field.Invalid(field.NewPath("spec").Child("runtimeRef"),
+					utiltesting.MakeTrainJobWrapper("default", "test").Obj().Spec.RuntimeRef,
+					fmt.Sprintf("must have %s job when trainJob is configured with input modelConfig", constants.DatasetInitializer)),
+			},
+		},
+		// assert that we get a error  here
+		"dataset initializer job exists but container missing": {
+			runtimeJobTemplate: func() client.Object {
+				js := utiltesting.MakeJobSetWrapper(metav1.NamespaceDefault, "test")
+				for i, rJob := range js.Spec.ReplicatedJobs {
+					if rJob.Name == constants.DatasetInitializer {
+						js.Spec.ReplicatedJobs[i].Template.Spec.Template.Spec.Containers = []v1.Container{}
+					}
+				}
+				return js.Obj()
+			}(),
+			newObj: utiltesting.MakeTrainJobWrapper("default", "test").
+				Initializer(&trainer.Initializer{
+					Dataset: &trainer.DatasetInitializer{},
+				}).Obj(),
+			wantError: field.ErrorList{
+				field.Invalid(field.NewPath("spec").Child("runtimeRef"),
+					utiltesting.MakeTrainJobWrapper("default", "test").Obj().Spec.RuntimeRef,
+					fmt.Sprintf("must have container with name - %s in the %s job", constants.DatasetInitializer, constants.DatasetInitializer)),
+			},
+		},
+		"no model intializer job": {
+			runtimeJobTemplate: utiltesting.MakeJobSetWrapper(metav1.NamespaceDefault, "test"),
+			newObj: utiltesting.MakeTrainJobWrapper(metav1.NamespaceDefault, "test").
+				Initializer(&trainer.Initializer{Model: nil}).
+				Obj(),
+			wantError: field.ErrorList{
+				field.Invalid(field.NewPath("spec").Child("runtimeRef"),
+					utiltesting.MakeTrainJobWrapper("default", "test").Obj().Spec.RuntimeRef,
+					fmt.Sprintf("must have %s job when trainJob is configured with input modelConfig", constants.ModelInitializer)),
+			},
+		},
+		"model intializer job exists but container missing ": {
+			runtimeJobTemplate: func() client.Object {
```

Comment:
Ditto, you can remove it because we don't need `runtimeJobTemplate` now.

- https://github.com/kubeflow/trainer/blob/4e6199c9486d861655a712d7017b8f23f9f2e48e/pkg/runtime/framework/plugins/jobset/jobset.go#L82